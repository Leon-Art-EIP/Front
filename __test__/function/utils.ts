import type { ReactTestRendererJSON } from "react-test-renderer";
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from "react-test-renderer";
import { expect } from "vitest";

function toJson(component: renderer.ReactTestRenderer): ReactTestRendererJSON {
  const result = component.toJSON();
  expect(result).toBeDefined();

  return result as ReactTestRendererJSON;
}

export function testSnapshot(element: JSX.Element): void {
  const component = renderer.create(element);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
}
