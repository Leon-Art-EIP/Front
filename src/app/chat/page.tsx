import ChatWrapper from "../../wrappers/chat/ChatWrapper";

export default function Page(): JSX.Element {
  return (
    <ChatWrapper></ChatWrapper>
  );
}
