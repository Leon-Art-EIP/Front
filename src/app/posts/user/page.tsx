import PostsWrapper from "../../../wrappers/posts/PostsWrapper";

export default function Page(): JSX.Element {
  return (
    <div className="flex">
      <div className="flex-1" />
      <div className="flex flex-col pt-4 gap-4">
        <h1>Posts</h1>
        <PostsWrapper filter="user" />
      </div>
      <div className="flex-1" />
    </div>
  );
}
