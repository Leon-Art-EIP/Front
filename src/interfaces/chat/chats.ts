export interface IChats {
  conversations: IChat[];
}

export interface IChat {
  id: number;
  profileName: string;
  profilePricture: string;
  lastMessage: string;
  unreadMessages: boolean;
}
