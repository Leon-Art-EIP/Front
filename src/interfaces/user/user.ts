export interface IUser {
  id: string;
  username: string;
  email: string;
  is_artist: boolean;
  availability: string;
  subscription: string;
  collections: string[];
}

export interface IConnectedUser {
  token: string;
  user: IUser;
}
