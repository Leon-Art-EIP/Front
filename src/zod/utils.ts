export const getValueOrUndefined = (arg: unknown) =>
  arg === "" || arg === undefined || arg === null ? undefined : arg;
