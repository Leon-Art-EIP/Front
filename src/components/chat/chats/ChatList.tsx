import { useState, useEffect } from "react";
import { useChat } from "../../../contexts/ChatContext";
import { IChat } from "../../../interfaces/chat/chats";
import { SearchBar } from "../../searchBar/SearchBar";
import { ChatUserCard } from "./ChatUserCard";
import { useRouter } from "next/navigation";

interface IChatListProps {
  onDeleteChat: () => void;
}

export default function ChatList(props: IChatListProps): JSX.Element {
  /* c8 ignore start */
  const router = useRouter();
  const { chats, currentUser, currentChat, setCurrentChat, refreshChats } = useChat() || {};
  const [searchTerm, setSearchTerm] = useState("");
  const filteredChats = filterChats(searchTerm);

  useEffect(() => {
    filterChats(searchTerm);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [chats, searchTerm]);

  function handleSearch(search: string) {
    setSearchTerm(search.toLowerCase());
  }

  function handleSelectChat(chat: IChat) {
    setCurrentChat(chat);
  }

  function filterChats(search: string) {
    const deletedChats = JSON.parse(localStorage.getItem("deletedChats") || "{}");

    const newFilteredChats = chats?.filter(
      (chat) =>
        (chat.UserOneName.toLowerCase().includes(search) || chat.UserTwoName.toLowerCase().includes(search)) &&
        !deletedChats[chat._id]
    );

    return newFilteredChats || [];
  }

  function handleDeleteChat(chatId: string) {
    refreshChats();
    props.onDeleteChat();
  }

  return (
    <div className="bg-background-hl flex flex-col h-full shadow-[2px_0_3px_0px_rgba(170,170,170)]">
      <div className="m-6 sm:m-10">
        <SearchBar onSearch={handleSearch} />
      </div>
      {filteredChats && filteredChats.length > 0 ? (
        <div className="grid grid-cols-1 gap-3">
          {filteredChats.map((chat, index) => (
            <ChatUserCard
              key={chat._id}
              chat={chat}
              currentUser={currentUser}
              handleSelectChat={handleSelectChat}
              currentSelected={currentChat}
              handleDeleteChat={handleDeleteChat}
            />
          ))}
        </div>
      ) : (
        <div className="flex justify-center w-full">
          <span className="text-sm italic">Aucune conversation</span>
        </div>
      )}
    </div>
  );
}
/* c8 ignore stop */
