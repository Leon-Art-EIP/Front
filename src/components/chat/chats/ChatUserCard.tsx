import { IChat } from "../../../interfaces/chat/chats";
import { IConnectedUser } from "../../../interfaces/user/user";
import { myFetch, NEXT_PUBLIC_BACKEND_URL } from "../../../tools/myFetch";
import MuteIcon from "@mui/icons-material/Notifications";
import UnmuteIcon from "@mui/icons-material/NotificationsOff";
import Delete from "@mui/icons-material/Delete";

import { useState } from "react";

export interface ChatUserCardProps {
  chat: IChat;
  currentUser: IConnectedUser | undefined;
  currentSelected: IChat | undefined;
  handleSelectChat: (chat: IChat) => void;
  handleDeleteChat: (chatId: string) => void;
}

export function ChatUserCard(props: ChatUserCardProps): JSX.Element {
  /* c8 ignore start */

  const [isMuted, setIsMuted] = useState<boolean>(() => {
    const mutedChats = JSON.parse(localStorage.getItem("mutedChats") || "{}");
    return mutedChats[props.chat._id] || false;
  });

  function onSelectChat() {
    props.handleSelectChat(props.chat);
  }

  function handleMuteToggle() {
    setIsMuted((prev) => {
      const newMutedState = !prev;

      const mutedChats = JSON.parse(localStorage.getItem("mutedChats") || "{}");
      mutedChats[props.chat._id] = newMutedState;
      localStorage.setItem("mutedChats", JSON.stringify(mutedChats));

      console.log("Mute/Unmute toggled", newMutedState);
      return newMutedState;
    });
  }

  async function handleDelete() {
    // const res = await myFetch({ route: `/api/conversations/delete/${props.chat._id}`, method: "DELETE" });
    // if (res.ok) {
    //   props.handleDeleteChat(props.chat._id);
    //   console.log("Chat deleted", props.chat._id);
    // } else {
    //   console.log("Failed to delete chat");
    // }
    const deletedChats = JSON.parse(localStorage.getItem("deletedChats") || "{}");
    deletedChats[props.chat._id] = true;
    localStorage.setItem("deletedChats", JSON.stringify(deletedChats));
    props.handleDeleteChat(props.chat._id);
    // console.log("Chat marked as deleted", props.chat._id);
  }

  return (
    <div className="pl-6 sm:pl-10">
      <div
        className={`relative flex flex-row flex-grow-0 w-full p-2 cursor-pointer rounded-l-full hover:bg-secondary group ${
          props.chat._id === props.currentSelected?._id ? "bg-secondary-hover" : ""
        }`}
        onClick={onSelectChat}
      >
        {props.chat.unreadMessages && !isMuted && props.chat.LastSenderId !== props.currentUser?.user.id && (
          <div className="absolute rounded-full w-3 h-3 bg-primary -left-4 top-1/2 transform -translate-y-1/2" />
        )}
        {(
          props.chat.UserOneId !== props.currentUser?.user.id ? props.chat.UserOnePicture : props.chat.UserTwoPicture
        ) ? (
          /* TODO: use <Image> next component */
          /* eslint-disable-next-line @next/next/no-img-element */
          <img
            src={`${NEXT_PUBLIC_BACKEND_URL}/api/${
              props.chat.UserOneId !== props.currentUser?.user.id
                ? props.chat.UserOnePicture
                : props.chat.UserTwoPicture
            }`}
            alt="profilePicture"
            className="w-16 h-16 rounded-full bg-secondary flex-shrink-0"
          />
        ) : (
          <span className="w-16 h-16 rounded-full bg-secondary flex-shrink-0"></span>
        )}
        <div className="flex flex-col justify-center flex-grow ml-4 text-tertiary">
          <span className="text-2xl tracking-wide truncate w-4/5">
            {props.chat.UserOneId !== props.currentUser?.user.id ? props.chat.UserOneName : props.chat.UserTwoName}
          </span>
          <div className="grid grid-cols-1">
            {props.chat._id !== props.currentSelected?._id ? (
              <span className="text-sm truncate">{props.chat.lastMessage}</span>
            ) : (
              <span className="text-md font-semibold">...</span>
            )}
          </div>
        </div>
        <button
          onClick={(e) => {
            e.stopPropagation();
            handleMuteToggle();
          }}
          className="absolute right-12 top-1/2 transform -translate-y-1/2 p-1 rounded-full bg-hover hidden group-hover:block"
        >
          {isMuted ? <UnmuteIcon /> : <MuteIcon />}
        </button>
        <button
          onClick={(e) => {
            e.stopPropagation();
            handleDelete();
          }}
          className="absolute right-4 top-1/2 transform -translate-y-1/2 p-1 rounded-full bg-hover hidden group-hover:block"
        >
          <Delete />
        </button>
        {props.chat._id === props.currentSelected?._id && (
          <div className="absolute right-0 top-0 w-1 h-full bg-primary" />
        )}
      </div>
    </div>
  );
}
/* c8 ignore stop */
