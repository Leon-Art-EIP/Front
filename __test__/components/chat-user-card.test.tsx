import { test } from "vitest";
import { testSnapshot } from "../function/utils";
import { ChatUserCard } from "../../src/components/chat/chats/ChatUserCard";
import { IChat } from "../../src/interfaces/chat/chats";
import { IConnectedUser } from "../../src/interfaces/user/user";

const chat: IChat = {
  _id: "1",
  lastMessage: "Hello",
  unreadMessages: true,
  UserOneId: "1",
  UserOneName: "User1",
  UserOnePicture: "user1.jpg",
  UserTwoId: "2",
  UserTwoName: "User2",
  UserTwoPicture: "user2.jpg",
  LastSenderId: "1",
};

const currentUser: IConnectedUser = {
  token: "token",
  user: {
    id: "1",
    username: "User1",
    email: "email",
    is_artist: false,
    availability: "available",
    subscription: "subscription",
    collections: [],
  },
};

test("ChatUserCard selected undefined user undefined snapshot", () => {
  testSnapshot(
    <ChatUserCard
      chat={chat}
      currentSelected={undefined}
      currentUser={undefined}
      handleDeleteChat={() => {}}
      handleSelectChat={() => {}}
    />
  );
});

test("ChatUserCard user undefined snapshot", () => {
  testSnapshot(
    <ChatUserCard
      chat={chat}
      currentSelected={chat}
      currentUser={undefined}
      handleDeleteChat={() => {}}
      handleSelectChat={() => {}}
    />
  );
});

test("ChatUserCard selected undefined snapshot", () => {
  testSnapshot(
    <ChatUserCard
      chat={chat}
      currentSelected={undefined}
      currentUser={currentUser}
      handleDeleteChat={() => {}}
      handleSelectChat={() => {}}
    />
  );
});

test("ChatUserCard everything defined snapshot", () => {
  testSnapshot(
    <ChatUserCard
      chat={chat}
      currentSelected={chat}
      currentUser={currentUser}
      handleDeleteChat={() => {}}
      handleSelectChat={() => {}}
    />
  );
});
