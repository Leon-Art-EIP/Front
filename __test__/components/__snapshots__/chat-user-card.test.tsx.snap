// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatUserCard everything defined snapshot 1`] = `
<div
  className="pl-6 sm:pl-10"
>
  <div
    className="relative flex flex-row flex-grow-0 w-full p-2 cursor-pointer rounded-l-full hover:bg-secondary group bg-secondary-hover"
    onClick={[Function]}
  >
    <img
      alt="profilePicture"
      className="w-16 h-16 rounded-full bg-secondary flex-shrink-0"
      src="undefined/api/user2.jpg"
    />
    <div
      className="flex flex-col justify-center flex-grow ml-4 text-tertiary"
    >
      <span
        className="text-2xl tracking-wide truncate w-4/5"
      >
        User2
      </span>
      <div
        className="grid grid-cols-1"
      >
        <span
          className="text-md font-semibold"
        >
          ...
        </span>
      </div>
    </div>
    <button
      className="absolute right-12 top-1/2 transform -translate-y-1/2 p-1 rounded-full bg-hover hidden group-hover:block"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="NotificationsIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 22c1.1 0 2-.9 2-2h-4c0 1.1.89 2 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z"
        />
      </svg>
    </button>
    <button
      className="absolute right-4 top-1/2 transform -translate-y-1/2 p-1 rounded-full bg-hover hidden group-hover:block"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </button>
    <div
      className="absolute right-0 top-0 w-1 h-full bg-primary"
    />
  </div>
</div>
`;

exports[`ChatUserCard selected undefined snapshot 1`] = `
<div
  className="pl-6 sm:pl-10"
>
  <div
    className="relative flex flex-row flex-grow-0 w-full p-2 cursor-pointer rounded-l-full hover:bg-secondary group "
    onClick={[Function]}
  >
    <img
      alt="profilePicture"
      className="w-16 h-16 rounded-full bg-secondary flex-shrink-0"
      src="undefined/api/user2.jpg"
    />
    <div
      className="flex flex-col justify-center flex-grow ml-4 text-tertiary"
    >
      <span
        className="text-2xl tracking-wide truncate w-4/5"
      >
        User2
      </span>
      <div
        className="grid grid-cols-1"
      >
        <span
          className="text-sm truncate"
        >
          Hello
        </span>
      </div>
    </div>
    <button
      className="absolute right-12 top-1/2 transform -translate-y-1/2 p-1 rounded-full bg-hover hidden group-hover:block"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="NotificationsIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 22c1.1 0 2-.9 2-2h-4c0 1.1.89 2 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z"
        />
      </svg>
    </button>
    <button
      className="absolute right-4 top-1/2 transform -translate-y-1/2 p-1 rounded-full bg-hover hidden group-hover:block"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`ChatUserCard selected undefined user undefined snapshot 1`] = `
<div
  className="pl-6 sm:pl-10"
>
  <div
    className="relative flex flex-row flex-grow-0 w-full p-2 cursor-pointer rounded-l-full hover:bg-secondary group "
    onClick={[Function]}
  >
    <div
      className="absolute rounded-full w-3 h-3 bg-primary -left-4 top-1/2 transform -translate-y-1/2"
    />
    <img
      alt="profilePicture"
      className="w-16 h-16 rounded-full bg-secondary flex-shrink-0"
      src="undefined/api/user1.jpg"
    />
    <div
      className="flex flex-col justify-center flex-grow ml-4 text-tertiary"
    >
      <span
        className="text-2xl tracking-wide truncate w-4/5"
      >
        User1
      </span>
      <div
        className="grid grid-cols-1"
      >
        <span
          className="text-sm truncate"
        >
          Hello
        </span>
      </div>
    </div>
    <button
      className="absolute right-12 top-1/2 transform -translate-y-1/2 p-1 rounded-full bg-hover hidden group-hover:block"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="NotificationsIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 22c1.1 0 2-.9 2-2h-4c0 1.1.89 2 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z"
        />
      </svg>
    </button>
    <button
      className="absolute right-4 top-1/2 transform -translate-y-1/2 p-1 rounded-full bg-hover hidden group-hover:block"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`ChatUserCard user undefined snapshot 1`] = `
<div
  className="pl-6 sm:pl-10"
>
  <div
    className="relative flex flex-row flex-grow-0 w-full p-2 cursor-pointer rounded-l-full hover:bg-secondary group bg-secondary-hover"
    onClick={[Function]}
  >
    <div
      className="absolute rounded-full w-3 h-3 bg-primary -left-4 top-1/2 transform -translate-y-1/2"
    />
    <img
      alt="profilePicture"
      className="w-16 h-16 rounded-full bg-secondary flex-shrink-0"
      src="undefined/api/user1.jpg"
    />
    <div
      className="flex flex-col justify-center flex-grow ml-4 text-tertiary"
    >
      <span
        className="text-2xl tracking-wide truncate w-4/5"
      >
        User1
      </span>
      <div
        className="grid grid-cols-1"
      >
        <span
          className="text-md font-semibold"
        >
          ...
        </span>
      </div>
    </div>
    <button
      className="absolute right-12 top-1/2 transform -translate-y-1/2 p-1 rounded-full bg-hover hidden group-hover:block"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="NotificationsIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 22c1.1 0 2-.9 2-2h-4c0 1.1.89 2 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z"
        />
      </svg>
    </button>
    <button
      className="absolute right-4 top-1/2 transform -translate-y-1/2 p-1 rounded-full bg-hover hidden group-hover:block"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </button>
    <div
      className="absolute right-0 top-0 w-1 h-full bg-primary"
    />
  </div>
</div>
`;
